File: frontend/src/api/stats.ts
import { apiRequest } from './client'

export type TopItemsTimeRange =
  | 'short_term'
  | 'medium_term'
  | 'long_term'

export interface TopArtist {
  rank: number
  spotify_id: string
  name: string
  image_url: string | null
  genres: string[]
  spotify_url: string | null
}

export interface TopArtistsResponse {
  time_range: TopItemsTimeRange
  limit: number
  total: number
  items: TopArtist[]
}

export function getTopArtists(
  timeRange: TopItemsTimeRange = 'medium_term',
  limit = 20,
): Promise<TopArtistsResponse> {
  const query = new URLSearchParams({
    time_range: timeRange,
    limit: String(limit),
  })

  return apiRequest<TopArtistsResponse>(
    `/api/stats/top-artists?${query.toString()}`,
  )
}

export interface TopTrack {
  rank: number
  spotify_id: string
  name: string
  artist_names: string[]
  album_name: string
  album_image_url: string | null
  duration_ms: number
  spotify_url: string | null
}

export interface TopTracksResponse {
  time_range: TopItemsTimeRange
  limit: number
  total: number
  items: TopTrack[]
}

export function getTopTracks(
  timeRange: TopItemsTimeRange = 'medium_term',
  limit = 20,
): Promise<TopTracksResponse> {
  const query = new URLSearchParams({
    time_range: timeRange,
    limit: String(limit),
  })

  return apiRequest<TopTracksResponse>(
    `/api/stats/top-tracks?${query.toString()}`,
  )
}

File: backend/app/schemas/stats.py
from typing import Literal

from pydantic import BaseModel, Field


TopItemsTimeRange = Literal[
    "short_term",
    "medium_term",
    "long_term",
]


class TopArtistResponse(BaseModel):
    rank: int
    spotify_id: str
    name: str
    image_url: str | None = None
    genres: list[str] = Field(default_factory=list)
    spotify_url: str | None = None


class TopArtistsResponse(BaseModel):
    time_range: TopItemsTimeRange
    limit: int
    total: int
    items: list[TopArtistResponse] = Field(default_factory=list)